File: src/main/kotlin/cn/bakamc/folia/db/table/PlayerInfo.kt
package cn.bakamc.folia.db.table

import org.ktorm.database.Database
import org.ktorm.entity.Entity
import org.ktorm.entity.sequenceOf
import org.ktorm.schema.Table
import org.ktorm.schema.varchar

open class PlayerInfos(alias: String?) : Table<PlayerInfo>("player", alias) {
    companion object : PlayerInfos(null)

    override fun aliased(alias: String): Table<PlayerInfo> = PlayerInfos(alias)

    val uuid = varchar("uuid").primaryKey().bindTo { it.uuid }
    val name = varchar("name").bindTo { it.name }
}

interface PlayerInfo : Entity<PlayerInfo> {
    companion object : Entity.Factory<PlayerInfo>()

    var uuid: String

    var name: String

}

val Database.playerInfos get() = this.sequenceOf(PlayerInfos)

File: bakamc-folia/src/main/scala/cn/bakamc/folia/database/table/PlayerInfo.scala
package cn.bakamc.folia.database.table

import org.bukkit.entity.Player
import slick.jdbc.MySQLProfile.api.*

import java.util.UUID
import scala.language.implicitConversions

object PlayerInfo {

  implicit def playerToInfo(player: Player): PlayerInfo = PlayerInfo(player)

  implicit def uuidToString(uuid: UUID): String = uuid.toString

  def apply(player: Player): PlayerInfo = {
    PlayerInfo(
      player.getUniqueId,
      player.getName
    )
  }

}

case class PlayerInfo(
  uuid: String,
  name: String,
)

object PlayerInfoImpl {
  given Conversion[Player, PlayerInfo] = PlayerInfo(_)
}


class PlayerInfos(tag: Tag) extends Table[PlayerInfo](tag, "player_info") {
  def uuid = column[String]("uuid", O.PrimaryKey, O.Length(36))

  def name = column[String]("name", O.Length(16))

  def * = (uuid, name).mapTo[PlayerInfo]
}

val playerInfos = TableQuery[PlayerInfos]
